### PR TITLE
Allow for web UI endpoints to be disabled during build

### PR DESCRIPTION
**Pull request by Blackburn29**
## Description
- Adds build flags and a new base configuration to _optionally_ enable the React Web UI. 

When working with Heltec V3 boards, memory pressure is tight for HTTP requests memory due to the lack of PSRAM. The webserver is disabled by default but I figured it would be useful to allow users to be able to utilize the API endpoints without the need of the dedicated UI. 

(In my use-case I am building a standalone react app that can be self-hosted to manage many NightDriverStrip devices and I need API access but do not need the embedded UI)

I welcome any feedback on the platform IO build configuration as it is still pretty new to me and I want to make sure that I do not increase the build complexity too much

## Contributing requirements

* [X] I read the contribution guidelines in [CONTRIBUTING.md](http://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/CONTRIBUTING.md).
* [X] I understand the BlinkenPerBit metric, and maximized it in this PR.
* [X] I selected `main` as the target branch.
* [X] All code herein is subjected to the license terms in [COPYING.txt](http://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/COPYING.txt).

### CODE REVIEW

**Comment by rbergen** [comment]
@Blackburn29 I agree with the premise of @robertlipe's feedback, but as far as I'm concerned it can be implemented with the `ENABLE_` approach. That could be done by defining `ENABLE_WEB_UI` to 1 at the bottom of globals.h, if it's not yet defined _and_ `ENABLE_WEBSERVER` is 1. 

So again, I do agree that "web UI enabled" needs to be the default when the webserver itself is enabled. With that in place, only configurations that do want the webserver but not the web UI need to be changed to define `ENABLE_WEB_UI` to 0.

**Comment by Blackburn29** [comment]
> @Blackburn29 I agree with the premise of @robertlipe's feedback, but as far as I'm concerned it can be implemented with the `ENABLE_` approach. That could be done by defining `ENABLE_WEB_UI` to 1 at the bottom of globals.h, if it's not yet defined _and_ `ENABLE_WEBSERVER` is 1.
> 
> So again, I do agree that "web UI enabled" needs to be the default when the webserver itself is enabled. With that in place, only configurations that do want the webserver but not the web UI need to be changed to define `ENABLE_WEB_UI` to 0.

How do you feel about the configuration I added in the platform config? Are you fine with that? It does add a level to complexity to the feature flag since you would need to change the base configuration of the device you are targeting in order to add/remove the embedded web files

**Comment by rbergen** [comment]
@Blackburn29 To be honest I hadn't gotten there yet, but now that you ask: no, let's not do that. We've been including the webserver files in builds without the webserver enabled the whole time, and I think we should stick to that. With the gzip compression in place we're talking about ~23K of flash memory, and in my opinion we can do without saving that. It's RAM where we run into limitations at the moment, not flash.  
Obviously, we can reconsider this if and when builds start failing over lack of flash space.

**Comment by Blackburn29** [comment]
@rbergen  I’m good with that. I’ll update the PR shortly

**Comment by Blackburn29** [comment]
Also, here is the standalone app if anyone is interested. 

image

https://github.com/Blackburn29/NightDriverWeb

**Comment by rbergen** [comment]
@Blackburn29 Is my understanding correct that you're still intending to push an update commit to reverse the changes to platformio.ini this PR now includes?

**Comment by Blackburn29** [comment]
> @Blackburn29 Is my understanding correct that you still intend to push an update commit to reverse the changes to platformio.ini that this PR now includes?

Yes, apologies. I’ve been busy at work and haven’t had a chance to make the changes. Hoping to get to it by the weekend

**Comment by rbergen** [comment]
@Blackburn29 Not a problem. I'm well familiar with the cadence. :)

**Comment by Blackburn29** [comment]
@rbergen The PR has been updated. This _should_ meet the requirements for what you were asking for.

File: include/globals.h [hunk 1]
```diff
@@ -1171,7 +1171,15 @@ extern RemoteDebug Debug;           // Let everyone in the project know about it
 #endif
 
 #ifndef ENABLE_WEBSERVER
-#define ENABLE_WEBSERVER        0   // Chip provides a web server with controls to adjust effects
+    #define ENABLE_WEBSERVER        0   // Chip provides a web server with controls to adjust effects
+
```
Comment (rbergen) [inline include/globals.h:1175]: I suggest we just leave this line alone and decrease the diff size a bit. We're not a big fan of whitespace-only changes, and this one breaks the "same indentation #ifndef/#define/#endif" pattern as well.

File: platformio.ini [hunk 2]
```diff
@@ -30,7 +30,8 @@ extra_configs =
 ; Options that are used (or extended) in all device sections (and hence environments) are defined here
 
 [base]
-upload_port     =
+#upload_protocol = 
+#upload_port     =
```
Comment (rbergen) [inline platformio.ini:34]: Please revert these changes as well.

File: platformio.ini [hunk 3]
```diff
@@ -245,11 +245,15 @@ lib_extra_dirs  = ${PROJECT_DIR}/lib
...
                   post:tools/merge_image.py
+
 board_build.embed_files = site/dist/index.html.gz
                           site/dist/index.js.gz
                           site/dist/favicon.ico.gz
+
 board_build.embed_txtfiles = config/timezones.json
 
+
+
```
Comment (rbergen) [inline platformio.ini:256]: Same here for the extra white lines.

File: src/webserver.cpp [hunk 4]
```diff
@@ -134,19 +134,17 @@ void CWebServer::begin()
...
     // Instance handler requests
-
     _server.on("/statistics",            HTTP_GET,  [this](AsyncWebServerRequest* pRequest) { this->GetStatistics(pRequest); });
     _server.on("/getStatistics",         HTTP_GET,  [this](AsyncWebServerRequest* pRequest) { this->GetStatistics(pRequest); });
 
     // Static handler requests
-
     _server.on("/effects",               HTTP_GET,  GetEffectListText);
     _server.on("/getEffectList",         HTTP_GET,  GetEffectListText);
     _server.on("/nextEffect",            HTTP_POST, NextEffect);
```
Comment (rbergen) [inline src/webserver.cpp:150]: And again.

**Review by rbergen — COMMENTED** [review]
@Blackburn29 Thanks for the update, but I do still have a few comments. They all boil down to one thing: please keep your changes limited to actual code changes, and leave unrelated things (like white spacing in different code areas) as they are.

**Comment by rbergen** [comment]
I've taken the liberty of applying my feedback myself. If CI completes successfully I will be happy to merge.
